### PR TITLE
[ci] add job for interval

### PR DESCRIPTION
**Pull request by gares**
because it's in ci-basic-overlay and is part of the windows installer

### CODE REVIEW

**Comment by gares** [comment]
@silene would you mind applying this fix https://gitlab.inria.fr/coqinterval/interval/-/issues/3

**Comment by gares** [comment]
dear @coq/ci-maintainers I need some help here, I did split ci-mathcomp in 3 jobs which are shorter, so that ci-interval only needs the first one, but something about artifacts does not work...

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -784,6 +784,36 @@ library:ci-fiat_crypto_ocaml:
...
+  needs:
+  - library:ci-coquelicot
+  - library:ci-flocq
+  - library:ci-mathcomp
+  - plugin:ci-bignums
+  dependencies:
+  - library:ci-coquelicot
+  - library:ci-flocq
+  - library:ci-mathcomp
+  - plugin:ci-bignums
```
Comment (SkySkimmer) [inline .gitlab-ci.yml:799]: You need the transitive needs and dependencies, ie including `build:edge+flambda`

Comment (gares) [reply .gitlab-ci.yml]: Thanks!

BTW, some jobs depend on build:base, some on build:edge+flambda ... any reason to pick one instead of the other?

Comment (SkySkimmer) [reply .gitlab-ci.yml]: Although considering https://github.com/coq/coq/pull/13471 I guess only `needs` is needed

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -784,6 +784,38 @@ library:ci-fiat_crypto_ocaml:
...
+
+library:ci-oddorder:
+  extends: .ci-template-flambda
+  stage: stage-3
+  needs:
+  - library:ci-mathcomp
+  dependencies:
+  - library:ci-mathcomp
```
Comment (SkySkimmer) [inline .gitlab-ci.yml:809]: This and fourcolor also need the transitive deps

**Comment by SkySkimmer** [comment]
The coquelicot script needs to call install.
The interval job seems to be rebuilding flocq for some reason, can remake not tell that it's up to date?

**Comment by silene** [comment]
> The interval job seems to be rebuilding flocq for some reason, can remake not tell that it's up to date?

Since the job fails at the `configure` step (if I understand correctly), `remake` has not even been called at that point. So, whatever is rebuilding Flocq, it is something else.

**Comment by SkySkimmer** [comment]
The configure for fourcolor is called after we call remake on the already-built flocq

**Comment by silene** [comment]
> The configure for fourcolor is called after we call remake on the already-built flocq

Did you mean Interval instead of Fourcolor? Otherwise I am completely lost.

Anyway, who is calling `remake`? It is not the one from `ci-interval.sh`, since the `configure` script for Interval has not yet completed.

**Comment by SkySkimmer** [comment]
>Did you mean Interval instead of Fourcolor? Otherwise I am completely lost.

Yes, sorry

>Anyway, who is calling remake? It is not the one from ci-interval.sh, since the configure script for Interval has not yet completed.

The one from ci-flocq.sh
CI works by calling `make -f Makefile.ci ci-foo`
Makefile.ci has all ci-foo as PHONY with a rule calling the ci-foo.sh
It also has the dependency information, eg `ci-interval: ci-flocq`
Therefore the job for interval first calls ci-flocq.sh, relying on the submake detecting that there is nothing to do.

File: dev/ci/ci-interval.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+ci_dir="$(dirname "$0")"
+. "${ci_dir}/ci-common.sh"
+
+git_download interval
+
+( cd "${CI_BUILD_DIR}/interval" && autoconf && ./configure && ./remake "-j${NJOBS}" && ./remake install )
```
Comment (gares) [inline dev/ci/ci-interval.sh:8]: @silene these kinds of lines are run multiple times, I guess I need a test like
`if [ ! -x ./configure ]; then autoconf && ./condifure; fi`

Comment (gares) [reply dev/ci/ci-interval.sh:8]: would this one work IYO?

**Comment by gares** [comment]
There is something wrong with coquelicot:
- it recompile a part of mathcomp
- it runs `./remake install` but later interval does not find it. is it installed in the wrong place? or is the artifact business not done properly?

File: dev/ci/ci-interval.sh (hunk 3 above)
Comment (silene) [reply dev/ci/ci-interval.sh:8]: That would be a solution.

In theory, `remake` is no different than `make`. That is, it does not recompile anything if the last-modified dates are properly ordered. So, there might be an issue with the way artifacts are handled, e.g., if two slaves have desynchronized clocks. But more probably, the issue just comes from the fact that the `configure` script of Flocq generates some of the `.v` files (due to compatibility issues across various Coq versions), which in turns forces `remake` to recompile their dependencies.

**Comment by silene** [comment]
> is it installed in the wrong place? or is the artifact business not done properly?

Flocq and Coquelicot use the same installation mechanism, so that would point to an issue with the artifacts (since Flocq succeeded).

**Comment by SkySkimmer** [comment]
Note that the artifacts contain the compilation directory but not the install directory, so rerunning the install command is important.

**Comment by SkySkimmer** [comment]
@gares does `make -f Makefile.ci ci-interval` work when you run it locally? (having prealably compiled coq either with dune or make)

**Comment by gares** [comment]
I think I tried it at some point. I'll do it again.

**Comment by gares** [comment]
Yes, it works in the sense that then if finds coquelicot. It still fails because of https://gitlab.inria.fr/coqinterval/interval/-/issues/3 not being merged yet (please @coq/build-maintainers 

I'm using make. So there is something wrong with the gitlab specific code I'm afraid.

**Comment by gares** [comment]
@SkySkimmer does last commit ring a bell? is it right?

**Comment by ejgallego** [comment]
Same problem with gitlab.inria and overlays.

File: dev/ci/ci-coquelicot.sh [hunk 4]
```diff
@@ -3,8 +3,6 @@
 ci_dir="$(dirname "$0")"
 . "${ci_dir}/ci-common.sh"
 
-install_ssreflect
```
Comment (SkySkimmer) [inline dev/ci/ci-coquelicot.sh:6]: This is wrong, coquelicot depends on ssr. Not sure what makes you think it's already installed.

Comment (gares) [reply dev/ci/ci-coquelicot.sh:6]: So, should I add `install_foo` in all jobs that depend on `foo`?

Comment (SkySkimmer) [reply dev/ci/ci-coquelicot.sh:6]: No, ssreflect is treated specially (because it's part of mathcomp but we don't want all mathcomp here)

**Comment by SkySkimmer** [comment]
I tried reading the log at https://gitlab.com/coq/coq/-/jobs/871864903 but interleaving is hard to parse (especially with this remake BS).
Could you remove the last commit and set NJOBS: 1 in the job for interval?

File: dev/ci/ci-coquelicot.sh (hunk 4 above)
Comment (gares) [reply dev/ci/ci-coquelicot.sh:6]: I'd like to remove this "exception", now coquelicot depends on mathcomp proper.
if you want, we can split mathcomp in 2.

**Comment by gares** [comment]
done, let's see. I still pick the flambda variant randomly, please provide some guidance

**Comment by SkySkimmer** [comment]
dependencies need to be in the Makefile.ci too
anyway just drop the install_ssr change so we have only 1 thing to debug

**Comment by gares** [comment]
```
Fatal error: exception Stack overflow
Error: Native compiler exited with status 2
```
I guess I can set the env flag to disable native as other jobs do. CC @silene

**Comment by silene** [comment]
> I guess I can set the env flag to disable native as other jobs do.

That is quite surprising. Is the CI enforcing an unusually small stack? Because the failing file is specifically designed to be compiled to native code. Anyway, to disable native compilation, just change the `./configure` call into

    ./configure COQEXTRAFLAGS="-native-compiler no"

**Comment by gares** [comment]
> Anyway, to disable native compilation, just change the `./configure` call into

Thanks I'll do that for now, I can't solve all the problems in one go.
As of today, these are the jobs that are in the same situation
```
dev/ci/ci-bedrock2.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-color.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-coqprime.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-corn.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-engine_bench.sh:export COQEXTRAFLAGS='-native-compiler ondemand'
dev/ci/ci-equations.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-fiat_crypto.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-fiat_crypto_legacy.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-fiat_crypto_ocaml.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-fiat_parsers.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-interval.sh:( cd "${CI_BUILD_DIR}/interval" && ( if [ ! -x ./configure COQEXTRAFLAGS="-native-compiler no" ]; then autoconf && ./configure; fi ) && ./remake "-j${NJOBS}" && ./remake install )
dev/ci/ci-metacoq.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-perennial.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-rewriter.sh:export COQEXTRAFLAGS='-native-compiler no'
dev/ci/ci-unimath.sh:export COQEXTRAFLAGS='-native-compiler no'
```

**Comment by silene** [comment]
By the way, you can use the exact same style as for the other jobs (I should have looked what they were doing before my suggestion):

     export COQEXTRAFLAGS='-native-compiler no'

will work just fine, as long as it is performed before `configure` is run.

**Comment by gares** [comment]
@SkySkimmer this is ready for review. It should fix some of the shortcoming of #13464

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -738,7 +738,11 @@ library:ci-coqtail:
   extends: .ci-template
 
 library:ci-coquelicot:
-  extends: .ci-template
+  stage: stage-3
+  extends: .ci-template-flambda
+  needs:
+  - build:edge+flambda
+  - library:ci-mathcomp
```
Comment (SkySkimmer) [inline .gitlab-ci.yml:745]: Because the template has `dependencies` you need to duplicate them here too.
I think this is why the coquelicot job rebuilt mathcomp.

File: dev/ci/ci-fourcolor.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+# $0 is not the safest way, but...
```
Comment (SkySkimmer) [inline dev/ci/ci-fourcolor.sh:3]: what's this comment for?

Comment (gares) [reply dev/ci/ci-fourcolor.sh:3]: I've no idea where it comes from

Comment (gares) [reply dev/ci/ci-fourcolor.sh:3]: it comes from ci-mathcomp.sh... I'll let you dig

File: .gitlab-ci.yml (hunk 5 above)
Comment (gares) [reply .gitlab-ci.yml:745]: I hope I fixed it

**Comment by SkySkimmer** [comment]
Why is there an appveyor check? oO
cc @Zimmi48

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Comment by gares** [comment]
great!

**Comment by Zimmi48** [comment]
For some reason AppVeyor was still installed on the Coq repo. I've reverted that.
